### ISSUE TITLE
clear screen Cmd

### ISSUE THREAD

**Issue by golonzovsky** [charmbracelet/bubbletea#236]
I would like to clear top of screen from within my app (same as Ctrl+L in bash). 

It seems I need to print "\033[H\033[2J" (for nix) and initiate full redraw. I see internal `renderer.repaint` is only initiated from `WindowSizeMsg`:
```go
		case "ctrl+l":
			fmt.Print("\033[H\033[2J") // only nix
			w, h, _ := term.GetSize(0) // wouldnt work if out is set
			return m, func() tea.Msg { // hack to trigger redraw
				return tea.WindowSizeMsg{Width: w, Height: h}
			}
		}
```

Is there an easier way to do this? 
Would be nice to have built-in `tea.ClearScreen` cmd or at least some way to trigger full redraw so solution is less hacky and cross-platform. 

![clear_screen](https://user-images.githubusercontent.com/2477950/154841815-01a4c02f-3c06-49b1-a4de-f9351aeb883d.gif)

**Comment by meowgorithm** [comment]
That’s an interesting idea. And thanks for the detailed explanation and GIF, all very helpful.

Just to make sure, are you trying to force a redraw of your application or clear the entire contents of the terminal?

Also, are you aware of the altscreen ([`tea.EnterAltScreen`](https://pkg.go.dev/github.com/charmbracelet/bubbletea#EnterAltScreen) and [`tea.WithAltScreen`](https://pkg.go.dev/github.com/charmbracelet/bubbletea#WithAltScreen)) which will clear the screen, print the output of your application, and restore the former terminal state on exit (similar to how Vim works)?

**Comment by golonzovsky** [comment]
I'm trying to do both clear and redraw, so when my app is exiting I have useful state on top of my term window. I just caught myself often trying to hit "ctrl+l" while in inline app, similar to this:
![ctrlL](https://user-images.githubusercontent.com/2477950/154867777-fd6df73e-f7f5-46cf-a488-5ffcf84c9adf.gif)

AltScreen will look same during app use, but after exit my last view will be in the middle of screen with all my typos from before:
![altScreen3](https://user-images.githubusercontent.com/2477950/154867497-673acc60-d2e6-455d-b790-1dea9f12268e.gif)

While I was looking at AltScreen code, I found how to easily clear screen: `termenv.ClearScreen()` 👍🏻 
So now I'm missing only redraw part.

**Comment by meowgorithm** [comment]
Okay, cool, I see your use case here. It _probably_ makes sense to add something like `tea.ClearScreen` as a command. It should really be done internally so it runs in-step with the renderer and given output (`termenv.ClearScreen` only operates on stdout, which is usually, but not always where Bubble Tea is writing output).

Side note for implementors: logic declares that there should never be a need to manually redraw the screen in a Bubble Tea application, so we likely won’t ever implement a screen clearing command.

**Comment by knz** [comment]
I've used the following, with success:

```
type doProgram func()

// Run is part of the tea.ExecCommand interface.
func (d doProgram) Run() error {
        d()
        return nil
}

// SetStdin is part of the tea.ExecCommand interface.
func (d doProgram) SetStdin(io.Reader) {}

// SetStdout is part of the tea.ExecCommand interface.
func (d doProgram) SetStdout(io.Writer) {}

// SetStderr is part of the tea.ExecCommand interface.
func (d doProgram) SetStderr(io.Writer) {}
```

then inside the `Update` code:

```
 return m, tea.Exec(doProgram(termenv.ClearScreen))
```

Arguably, this could be automated inside tea.go.

**Comment by knz** [comment]
@meowgorithm coming back to this:

> Side note for implementors: logic declares that there should never be a need to manually redraw the screen in a Bubble Tea application, so we likely won’t ever implement a screen clearing command.

While I agree about "there should never be a need to redraw the screen", note that the Ctrl+L use case has another part: it's also a way to move the cursor to the top left, to **gain more visual space to do things**. 

For example, we use this in [bubbline](https://github.com/knz/bubbline) to expand the viewing area.

Because of this legitimate use case, I think it's reasonable to integrate this feature as a *logical*  thing.

**Comment by knz** [comment]
It's also a way for the user to remove the visual clutter of whatever output was present on the screen above the bubble output, when the alt screen is not used.

**Comment by philpennock** [comment]
To record another use-case I believe to be legitimate (for people considering future cleanups): one tool I use has the ability to open a browser on a specific URL, but the Linux desktop's browser binding ends up spewing messages to the tty when it opens a browser.  This corrupts the display and there's no sane way for the app to fully protect against this.  I contributed support for Ctrl-L binding last year; but with a rewrite to use Bubbletea v2, the mechanism I found disappeared; searching for another means, I found this issue and `tea.ClearScreen`.  It works perfectly.

I agree that it shouldn't be needed, but to recover from a messy real world, a recovery flow is necessary and `tea.ClearScreen` is the pragmatic solution.
